doc/contributing/py-2-3-guide.rst

```diff
@@ -0,0 +1,117 @@
+=====================
+CKAN Python 2/3 Guide
+=====================
+
+This section documents the requirements for porting existing CKAN code for Python 3 support, and for creating new code that supports both Python 2 and Python 3.
+
+----------
+User story
+----------
+
+As a developer working on CKAN or CKAN Extensions, I want a clear guide, for writing code compatible with Python 2 and Python 3, so my code is consistent with common practice around CKAN, using a common set of tools.
+
+---------
+Use cases
+---------
+
+- Contributing Python 2/3 compatible code in CKAN Core
+- Contributing Python 2/3 compatible code in a CKAN Extension
+- Starting a new CKAN Extension with Python 2/3 compatibility
+
+--------
+Python 3
+--------
+
+From CKAN 2.9, CKAN will officially support Python 2 (≥ 2.6) and Python 3 (≥ 3.6).
+
+-----
+Flask
+-----
+
+CKAN provides Python 3 support by removing the underlying Pylons web framework, which is only compatible with Python 2, and integrating Flask as the web framework, which is compatible with Python 2 and Python 3. Flask is a very popular framework for writing web applications in Python; read more about Flask_.
+
+--------
+Approach
+--------
+
+CKAN provides support for both Python 2 and Python 3 using a compatible source approach. This enables us to continue supporting Python 2 in the current transitional period, and gracefully upgrade to a Python 3 only codebase in future versions of CKAN (from CKAN 3.0 onwards).
+
+----
+Tips
+----
+
+Future imports
+##############
+
+Import the following at the top of all modules::
+
+  from __future__ import absolute_import
+  from __future__ import division
+  from __future__ import print_function
+  from __future__ import unicode_literals
+
+String handling
+###############
+
+- Remove any ``u`` prefixes before unicode strings in existing Python 2 code.
+- Add a ``b`` prefix for any bytestrings
+- Review code to ensure only unicode strings are exposed as part of the public interfaces of the codebase, which will make string handling more consistent.
+  
+Ensure you read the references below for further background information on all these tips.
+
+Exception handling
+##################
+
+Python-Future provides a clean interface for exception handling compatibility::
+
+  from future.utils import raise_
+  traceback = sys.exc_info()[2]
+  raise_(ValueError, "dodgy value", traceback)
```

Comment:
-1 on python-future. I strongly dislike rewriting exceptions this way. I think the cases in ckan core where we raise exceptions that won't work in python 2/3 should just be fixed (if there are any)

Reply:
Hey, thanks.

As far as I know, raising exceptions with a traceback, and also reprising exceptions, across both 2 and 3 is only possible like this. Of course the standard approach of `Exception as e` is better for simple cases.

Some more info [here](https://python-future.org/compatible_idioms.html#raising-exceptions) and [here](https://franklingu.github.io/programming/2016/06/30/properly-reraise-exception-in-python/).

Reply:
I agree this seems overkill. I grepped the code for exceptions that attach tracebacks and couldn't find any (maybe some on the search index rebuild CLI). There are several instances where we catch exceptions and we reraise custom ones like, but most are at the validators and logic level and are later catch by the blueprints.

I think is best to deal with problematic cases as we find them